# Проверка корректности данных

## Загрузка данных и подключение библиотек

In [2]:
#импорт необходимых библиотек
import pandas as pd
import numpy as np

#загрузка датасета
df = pd.read_csv('../data/flipkard.csv')
#создаем копию данных
df_clean = df.copy()
#проверяем, что данные успешно загружены
df_clean.head()

product_id      product_name        category   brand        seller  \
0  FKP0000001  Adidas Ultra 664            Toys  Adidas     MegaStore   
1  FKP0000002     LG Series 124         Fashion      LG     ValueKart   
2  FKP0000003    Redmi Model 35          Beauty   Redmi    SmartDeals   
3  FKP0000004  Sony Edition 769            Toys    Sony  UrbanRetails   
4  FKP0000005    Boat Prime 291  Home & Kitchen    Boat     RetailHub   

  seller_city     price  discount_percent  final_price  rating  ...  weight_g  \
0   Hyderabad  35547.34                15     30215.24     1.9  ...   3038.23   
1      Mumbai  30693.79                10     27624.41     3.2  ...   1921.82   
2   Ahmedabad  51214.50                40     30728.70     4.4  ...   1143.49   
3       Delhi  33168.49                30     23217.94     2.8  ...   3644.81   
4        Pune  14181.34                 0     14181.34     3.7  ...    310.33   

   warranty_months  color      size  return_policy_days  is_returnable  \
0               36  Green        XL                  30           True   
1                0   Grey  One Size                   7           True   
2               12  Black         M                   0          False   
3               24  Black         L                   0          False   
4                0   Blue       NaN                  30           True   

   payment_modes shipping_weight_g product_score  seller_rating  
0       COD,CARD       3483.592454          6.78           3.28  
1       UPI,CARD       2091.473835         50.78           3.42  
2   COD,UPI,CARD       1242.770494          4.24           4.23  
3    CARD,Wallet       3891.222716          1.38           3.45  
4    CARD,Wallet        386.061014         79.27           4.90  

[5 rows x 25 columns]

In [3]:
#размерность датасета
df_clean.shape

(80000, 25)

In [4]:
#названия колонок
df_clean.columns

Index(['product_id', 'product_name', 'category', 'brand', 'seller',
       'seller_city', 'price', 'discount_percent', 'final_price', 'rating',
       'review_count', 'stock_available', 'units_sold', 'listing_date',
       'delivery_days', 'weight_g', 'warranty_months', 'color', 'size',
       'return_policy_days', 'is_returnable', 'payment_modes',
       'shipping_weight_g', 'product_score', 'seller_rating'],
      dtype='object')

In [5]:
#общая инфомация о датасете
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80000 entries, 0 to 79999
Data columns (total 25 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_id          80000 non-null  object 
 1   product_name        80000 non-null  object 
 2   category            80000 non-null  object 
 3   brand               80000 non-null  object 
 4   seller              80000 non-null  object 
 5   seller_city         80000 non-null  object 
 6   price               80000 non-null  float64
 7   discount_percent    80000 non-null  int64  
 8   final_price         80000 non-null  float64
 9   rating              80000 non-null  float64
 10  review_count        80000 non-null  int64  
 11  stock_available     80000 non-null  int64  
 12  units_sold          80000 non-null  int64  
 13  listing_date        80000 non-null  object 
 14  delivery_days       80000 non-null  int64  
 15  weight_g            80000 non-null  float64
 16  warr

## Приведение типов данных

In [6]:
#просмотр типов данных
df_clean.dtypes

product_id             object
product_name           object
category               object
brand                  object
seller                 object
seller_city            object
price                 float64
discount_percent        int64
final_price           float64
rating                float64
review_count            int64
stock_available         int64
units_sold              int64
listing_date           object
delivery_days           int64
weight_g              float64
warranty_months         int64
color                  object
size                   object
return_policy_days      int64
is_returnable            bool
payment_modes          object
shipping_weight_g     float64
product_score         float64
seller_rating         float64
dtype: object

In [7]:
#просмотр уникальных значений
df_clean['listing_date'].unique()[:20]

array(['2022-04-12', '2023-05-14', '2022-09-13', '2019-08-05',
       '2018-08-26', '2018-04-15', '2021-03-17', '2020-04-08',
       '2018-01-14', '2022-05-15', '2018-02-28', '2022-03-09',
       '2020-06-02', '2023-03-16', '2020-02-17', '2022-08-23',
       '2018-08-27', '2020-04-15', '2022-12-05', '2023-06-18'],
      dtype=object)

In [8]:
#преобразование дат
df_clean['listing_date'] = pd.to_datetime(df['listing_date'], errors='coerce')
#убираем пропуски
df_clean = df_clean.dropna(subset=['listing_date'])

In [9]:
#проверяем пропуски
print(df_clean['listing_date'].isna().sum())

0


In [10]:
#преобразование категорий
category_names = ['category', 'brand', 'seller', 'seller_city', 'color', 'size', 'payment_modes']
for name in category_names:
    df[name] = df[name].astype('category')

In [11]:
#проверка результатов
df_clean.dtypes

product_id                    object
product_name                  object
category                      object
brand                         object
seller                        object
seller_city                   object
price                        float64
discount_percent               int64
final_price                  float64
rating                       float64
review_count                   int64
stock_available                int64
units_sold                     int64
listing_date          datetime64[ns]
delivery_days                  int64
weight_g                     float64
warranty_months                int64
color                         object
size                          object
return_policy_days             int64
is_returnable                   bool
payment_modes                 object
shipping_weight_g            float64
product_score                float64
seller_rating                float64
dtype: object

## Анализ пропущенных значений

In [12]:
#количество пропусков в каждом столбце
df_clean.isnull().sum()

product_id                0
product_name              0
category                  0
brand                     0
seller                    0
seller_city               0
price                     0
discount_percent          0
final_price               0
rating                    0
review_count              0
stock_available           0
units_sold                0
listing_date              0
delivery_days             0
weight_g                  0
warranty_months           0
color                     0
size                  13484
return_policy_days        0
is_returnable             0
payment_modes             0
shipping_weight_g         0
product_score             0
seller_rating             0
dtype: int64

In [13]:
#заполнение пропусков
df_clean['size'] = df_clean['size'].fillna('Unknown')

In [14]:
#доля пропусков в процентах
(df_clean.isnull().mean() * 100).round(2)

product_id            0.0
product_name          0.0
category              0.0
brand                 0.0
seller                0.0
seller_city           0.0
price                 0.0
discount_percent      0.0
final_price           0.0
rating                0.0
review_count          0.0
stock_available       0.0
units_sold            0.0
listing_date          0.0
delivery_days         0.0
weight_g              0.0
warranty_months       0.0
color                 0.0
size                  0.0
return_policy_days    0.0
is_returnable         0.0
payment_modes         0.0
shipping_weight_g     0.0
product_score         0.0
seller_rating         0.0
dtype: float64

## Проверка дубликатов

In [15]:
#количество дублирующихся строк
df_clean.duplicated().sum()

np.int64(0)

In [16]:
#удаление дубликатов при необходимости
df_clean = df_clean.drop_duplicates()

## Проверка диапазонов значений

In [17]:
#описательная статистика для числовых признаков
df_clean.describe()

price  discount_percent   final_price        rating  \
count  80000.000000      80000.000000  80000.000000  80000.000000   
mean   30137.256454         21.350625  23697.780600      2.999671   
min      200.140000          0.000000    101.080000      1.000000   
25%    15180.205000         10.000000  11341.722500      2.000000   
50%    30164.455000         20.000000  22571.265000      3.000000   
75%    45077.087500         40.000000  34521.057500      4.000000   
max    59995.800000         50.000000  59995.800000      5.000000   
std    17274.048682         16.354446  14726.821259      1.156238   

       review_count  stock_available    units_sold  \
count  80000.000000     80000.000000  80000.000000   
mean   25102.249875       499.194262   2507.515775   
min        0.000000         0.000000      0.000000   
25%    12605.000000       249.000000   1261.000000   
50%    25166.500000       499.000000   2513.000000   
75%    37666.000000       749.000000   3754.000000   
max    49999.000000       999.000000   4999.000000   
std    14428.769942       288.917706   1439.519972   

                     listing_date  delivery_days      weight_g  \
count                       80000   80000.000000  80000.000000   
mean   2020-09-28 12:16:35.760000       6.006325   2522.224622   
min           2018-01-01 00:00:00       1.000000     50.020000   
25%           2019-05-16 00:00:00       3.000000   1283.117500   
50%           2020-10-01 00:00:00       6.000000   2524.550000   
75%           2022-02-09 00:00:00       9.000000   3753.195000   
max           2023-06-24 00:00:00      11.000000   4999.920000   
std                           NaN       3.160630   1428.287966   

       warranty_months  return_policy_days  shipping_weight_g  product_score  \
count      80000.00000        80000.000000       80000.000000   80000.000000   
mean          15.56400           12.362250        2899.960047      50.673750   
min            0.00000            0.000000          53.155250       1.000000   
25%            6.00000            7.000000        1469.583781      25.950000   
50%           12.00000           10.000000        2892.632502      50.670000   
75%           24.00000           15.000000        4304.695597      75.430000   
max           36.00000           30.000000        6234.757448     100.000000   
std           12.91307           10.019794        1650.921766      28.562732   

       seller_rating  
count   80000.000000  
mean        4.002218  
min         3.000000  
25%         3.500000  
50%         4.010000  
75%         4.500000  
max         5.000000  
std         0.577595

In [18]:
#проверка логический границ
df_clean[df_clean['discount_percent'] > 100]л

SyntaxError: invalid syntax (2330920398.py, line 2)

## Проверка категориальных значений

In [ ]:
#частота уникальных значений
df_clean['category'].value_counts()

category
Toys              10151
Beauty            10094
Fashion           10080
Electronics       10052
Sports             9985
Appliances         9971
Mobiles            9904
Home & Kitchen     9763
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['brand'].value_counts()

brand
Puma         5409
Prestige     5401
Apple        5389
Reebok       5387
Nike         5358
LG           5345
Samsung      5345
Dell         5341
Redmi        5340
Philips      5340
HP           5297
Sony         5287
Adidas       5281
Boat         5258
Whirlpool    5222
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['seller'].value_counts()

seller
ValueKart       10109
SuperMart       10075
UrbanRetails    10067
MegaStore       10001
QuickShop        9985
BestBuy          9980
RetailHub        9940
SmartDeals       9843
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['seller_city'].value_counts()

seller_city
Pune         10138
Delhi        10130
Chennai      10019
Ahmedabad    10007
Mumbai        9986
Hyderabad     9956
Bengaluru     9944
Kolkata       9820
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['color'].value_counts()

color
Silver    10101
Black     10068
White     10034
Green     10001
Blue       9978
Grey       9969
Gold       9965
Red        9884
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['size'].value_counts()

size
Unknown     13484
L           13450
XL          13408
M           13317
One Size    13222
S           13119
Name: count, dtype: int64

In [ ]:
#частота уникальных значений
df_clean['payment_modes'].value_counts()

payment_modes
CARD,Wallet     20114
UPI,CARD        20037
COD,UPI,CARD    19977
COD,CARD        19872
Name: count, dtype: int64

## Логическая проверка взаимосвязей

In [ ]:
#поверка согласованности расчетных значений
mask = ~np.isclose(
    df_clean['final_price'],
    df_clean['price'] * (1 - df_clean['discount_percent'] / 100),
    atol=0.01)
df_clean[mask]

Empty DataFrame
Columns: [product_id, product_name, category, brand, seller, seller_city, price, discount_percent, final_price, rating, review_count, stock_available, units_sold, listing_date, delivery_days, weight_g, warranty_months, color, size, return_policy_days, is_returnable, payment_modes, shipping_weight_g, product_score, seller_rating]
Index: []

[0 rows x 25 columns]

## Проверка уникальности идентификаторов

In [ ]:
df_clean['product_id'].is_unique

True

## Результаты проверки

– Пропущенные значения: size: 13484 строк – заполнены 'Unknown'  
– Логические ошибки: final_price: 0 ошибок  
– Дубликаты: отсутсвуют   
– Типы данных: все столбцы приведены к корретным типам  


# Сохранение данных

In [ ]:
#сохраняем новый csv файл
df_clean.to_csv('../data/flipkard_cleaned.csv', index=False)